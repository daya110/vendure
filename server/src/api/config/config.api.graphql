type Query {
  config: Config!
}

type Config {
    customFields: JSON
}
