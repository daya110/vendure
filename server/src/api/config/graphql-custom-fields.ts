import { buildSchema, extendSchema, parse, printSchema } from 'graphql';

import { CustomFieldConfig, CustomFields, CustomFieldType } from '../../../../shared/shared-types';
import { assertNever } from '../../../../shared/shared-utils';

/**
 * Given a CustomFields config object, generates an SDL string extending the built-in
 * types with a customFields property for all entities, translations and inputs for which
 * custom fields are defined.
 */
export function addGraphQLCustomFields(typeDefs: string, customFieldConfig: CustomFields): string {
    const schema = buildSchema(typeDefs);

    let customFieldTypeDefs = '';

    if (!schema.getType('JSON')) {
        customFieldTypeDefs += `
            scalar JSON
        `;
    }

    for (const entityName of Object.keys(customFieldConfig)) {
        const customEntityFields = customFieldConfig[entityName as keyof CustomFields] || [];

        const localeStringFields = customEntityFields.filter(field => field.type === 'localeString');
        const nonLocaleStringFields = customEntityFields.filter(field => field.type !== 'localeString');

        if (schema.getType(entityName)) {
            if (customEntityFields.length) {
                customFieldTypeDefs += `
                    type ${entityName}CustomFields {
                        ${mapToFields(customEntityFields, getGraphQlType)}
                    }

                    extend type ${entityName} {
                        customFields: ${entityName}CustomFields
                    }
                `;
            } else {
                customFieldTypeDefs += `
                    extend type ${entityName} {
                        customFields: JSON
                    }
                `;
            }
        }

        if (localeStringFields.length && schema.getType(`${entityName}Translation`)) {
            customFieldTypeDefs += `
                    type ${entityName}TranslationCustomFields {
                         ${mapToFields(localeStringFields, getGraphQlType)}
                    }

                    extend type ${entityName}Translation {
                        customFields: ${entityName}TranslationCustomFields
                    }
                `;
        }

        if (schema.getType(`Create${entityName}Input`)) {
            if (nonLocaleStringFields.length) {
                customFieldTypeDefs += `
                    input Create${entityName}CustomFieldsInput {
                       ${mapToFields(nonLocaleStringFields, getGraphQlType)}
                    }

                    extend input Create${entityName}Input {
                        customFields: Create${entityName}CustomFieldsInput
                    }
                `;
            } else {
                customFieldTypeDefs += `
                   extend input Create${entityName}Input {
                       customFields: JSON
                   }
               `;
            }
        }

        if (schema.getType(`Update${entityName}Input`)) {
            if (nonLocaleStringFields.length) {
                customFieldTypeDefs += `
                    input Update${entityName}CustomFieldsInput {
                       ${mapToFields(nonLocaleStringFields, getGraphQlType)}
                    }

                    extend input Update${entityName}Input {
                        customFields: Update${entityName}CustomFieldsInput
                    }
                `;
            } else {
                customFieldTypeDefs += `
                    extend input Update${entityName}Input {
                        customFields: JSON
                    }
                `;
            }
        }

        if (customEntityFields.length && schema.getType(`${entityName}SortParameter`)) {
            customFieldTypeDefs += `
                    extend input ${entityName}SortParameter {
                         ${mapToFields(customEntityFields, () => 'SortOrder')}
                    }
                `;
        }

        if (customEntityFields.length && schema.getType(`${entityName}FilterParameter`)) {
            customFieldTypeDefs += `
                    extend input ${entityName}FilterParameter {
                         ${mapToFields(customEntityFields, getFilterOperator)}
                    }
                `;
        }

        if (localeStringFields && schema.getType(`${entityName}TranslationInput`)) {
            if (localeStringFields.length) {
                customFieldTypeDefs += `
                    input ${entityName}TranslationCustomFieldsInput {
                        ${mapToFields(localeStringFields, getGraphQlType)}
                    }

                    extend input ${entityName}TranslationInput {
                        customFields: ${entityName}TranslationCustomFieldsInput
                    }
                `;
            } else {
                customFieldTypeDefs += `
                    extend input ${entityName}TranslationInput {
                        customFields: JSON
                    }
                `;
            }
        }
    }

    return printSchema(extendSchema(schema, parse(customFieldTypeDefs)));
}

type GraphQLFieldType = 'String' | 'Int' | 'Float' | 'Boolean' | 'ID';

/**
 * Maps an array of CustomFieldConfig objects into a string of SDL fields.
 */
function mapToFields(fieldDefs: CustomFieldConfig[], typeFn: (fieldType: CustomFieldType) => string): string {
    return fieldDefs.map(field => `${field.name}: ${typeFn(field.type)}`).join('\n');
}

function getFilterOperator(type: CustomFieldType): string {
    switch (type) {
        case 'datetime':
            return 'DateOperators';
        case 'string':
        case 'localeString':
            return 'StringOperators';
        case 'boolean':
            return 'BooleanOperators';
        case 'int':
        case 'float':
            return 'NumberOperators';
        default:
            assertNever(type);
    }
    return 'String';
}

function getGraphQlType(type: CustomFieldType): GraphQLFieldType {
    switch (type) {
        case 'string':
        case 'datetime':
        case 'localeString':
            return 'String';
        case 'boolean':
            return 'Boolean';
        case 'int':
            return 'Int';
        case 'float':
            return 'Float';
        default:
            assertNever(type);
    }
    return 'String';
}
