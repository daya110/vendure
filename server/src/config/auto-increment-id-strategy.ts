import { IntegerIdStrategy } from './entity-id-strategy';

/**
 * An id strategy which simply uses auto-increment integers as primary keys
 * for all entities.
 */
export class AutoIncrementIdStrategy implements IntegerIdStrategy {
    primaryKeyType: 'increment';
    decodeId(id: string): number {
        return +id;
    }
    encodeId(primaryKey: number): string {
        return primaryKey.toString();
    }
}
