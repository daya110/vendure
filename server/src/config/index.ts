export * from './asset-naming-strategy/asset-naming-strategy';
export * from './asset-naming-strategy/default-asset-naming-strategy';
export * from './asset-preview-strategy/asset-preview-strategy';
export * from './asset-storage-strategy/asset-storage-strategy';
export * from './email/email-options';
export * from './email/email-transport-options';
export * from './entity-id-strategy/auto-increment-id-strategy';
export * from './entity-id-strategy/entity-id-strategy';
export * from './entity-id-strategy/uuid-id-strategy';
export * from './order-merge-strategy/order-merge-strategy';
export * from './payment-method/fakepal-payment-method-config';
export * from './payment-method/gripe-payment-method-config';
export * from './payment-method/payment-method-handler';
export * from './promotion/default-promotion-actions';
export * from './promotion/default-promotion-conditions';
export * from './promotion/promotion-action';
export * from './promotion/promotion-condition';
export * from './shipping-method/default-shipping-calculator';
export * from './shipping-method/default-shipping-eligibility-checker';
export * from './shipping-method/shipping-calculator';
export * from './shipping-method/shipping-eligibility-checker';
export * from './vendure-config';
export * from './vendure-plugin/vendure-plugin';
