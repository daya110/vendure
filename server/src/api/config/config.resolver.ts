import { Query, Resolver } from '@nestjs/graphql';

import { getConfig, VendureConfig } from '../../config/vendure-config';
import { ConfigService } from '../../service/config.service';

@Resolver('Config')
export class ConfigResolver {
    constructor(private configService: ConfigService) {}

    /**
     * Exposes a subset of the VendureConfig which may be of use to clients.
     */
    @Query()
    config(): Partial<VendureConfig> {
        return {
            customFields: this.configService.customFields,
        };
    }
}
