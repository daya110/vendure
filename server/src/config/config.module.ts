import { Module } from '@nestjs/common';

import { I18nService } from '../i18n/i18n.service';

import { ConfigService } from './config.service';

@Module({
    providers: [ConfigService],
    exports: [ConfigService],
})
export class ConfigModule {}
