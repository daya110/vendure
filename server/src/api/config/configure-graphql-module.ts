import { DynamicModule } from '@nestjs/common';
import { GqlModuleOptions, GraphQLModule, GraphQLTypesLoader } from '@nestjs/graphql';
import { GraphQLUpload } from 'apollo-server-core';
import { extendSchema, printSchema } from 'graphql';
import { GraphQLDateTime } from 'graphql-iso-date';
import GraphQLJSON from 'graphql-type-json';

import { ConfigModule } from '../../config/config.module';
import { ConfigService } from '../../config/config.service';
import { I18nModule } from '../../i18n/i18n.module';
import { I18nService } from '../../i18n/i18n.service';
import { getPluginAPIExtensions } from '../../plugin/plugin-utils';
import { ApiSharedModule } from '../api-internal-modules';
import { IdCodecService } from '../common/id-codec.service';
import { IdEncoderExtension } from '../middleware/id-encoder-extension';
import { TranslateErrorExtension } from '../middleware/translate-errors-extension';

import { generateListOptions } from './generate-list-options';
import { addGraphQLCustomFields } from './graphql-custom-fields';

export interface GraphQLApiOptions {
    apiType: 'shop' | 'admin';
    typePaths: string[];
    apiPath: string;
    // tslint:disable-next-line:ban-types
    resolverModule: Function;
}

/**
 * Dynamically generates a GraphQLModule according to the given config options.
 */
export function configureGraphQLModule(
    getOptions: (configService: ConfigService) => GraphQLApiOptions,
): DynamicModule {
    return GraphQLModule.forRootAsync({
        useFactory: (
            configService: ConfigService,
            i18nService: I18nService,
            idCodecService: IdCodecService,
            typesLoader: GraphQLTypesLoader,
        ) => {
            return createGraphQLOptions(
                i18nService,
                configService,
                idCodecService,
                typesLoader,
                getOptions(configService),
            );
        },
        inject: [ConfigService, I18nService, IdCodecService, GraphQLTypesLoader],
        imports: [ConfigModule, I18nModule, ApiSharedModule],
    });
}

function createGraphQLOptions(
    i18nService: I18nService,
    configService: ConfigService,
    idCodecService: IdCodecService,
    typesLoader: GraphQLTypesLoader,
    options: GraphQLApiOptions,
): GqlModuleOptions {
    // Prevent `Type "Node" is missing a "resolveType" resolver.` warnings.
    // See https://github.com/apollographql/apollo-server/issues/1075
    const dummyResolveType = {
        __resolveType() {
            return null;
        },
    };

    return {
        path: '/' + options.apiPath,
        typeDefs: createTypeDefs(options.apiType),
        include: [options.resolverModule],
        resolvers: {
            JSON: GraphQLJSON,
            DateTime: GraphQLDateTime,
            Node: dummyResolveType,
            PaginatedList: dummyResolveType,
            Upload: GraphQLUpload || dummyResolveType,
            SearchResultPrice: {
                __resolveType(value: any) {
                    return value.hasOwnProperty('value') ? 'SinglePrice' : 'PriceRange';
                },
            },
        },
        uploads: {
            maxFileSize: configService.assetOptions.uploadMaxFileSize,
        },
        playground: true,
        debug: true,
        context: (req: any) => req,
        extensions: [
            () => new TranslateErrorExtension(i18nService),
            () => new IdEncoderExtension(idCodecService),
        ],
        // This is handled by the Express cors plugin
        cors: false,
    };

    /**
     * Generates the server's GraphQL schema by combining:
     * 1. the default schema as defined in the source .graphql files specified by `typePaths`
     * 2. any custom fields defined in the config
     * 3. any schema extensions defined by plugins
     */
    function createTypeDefs(apiType: 'shop' | 'admin'): string {
        const customFields = configService.customFields;
        const typeDefs = typesLoader.mergeTypesByPaths(...options.typePaths);
        let schema = generateListOptions(typeDefs);
        schema = addGraphQLCustomFields(schema, customFields);
        const pluginSchemaExtensions = getPluginAPIExtensions(configService.plugins, apiType).map(
            e => e.schema,
        );

        for (const documentNode of pluginSchemaExtensions) {
            schema = extendSchema(schema, documentNode);
        }
        return printSchema(schema);
    }
}
