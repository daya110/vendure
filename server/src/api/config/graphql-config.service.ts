import { DynamicModule } from '@nestjs/common';
import { GqlModuleOptions, GraphQLModule, GraphQLTypesLoader } from '@nestjs/graphql';
import { GraphQLUpload } from 'apollo-server-core';
import { extendSchema, printSchema } from 'graphql';
import { GraphQLDateTime } from 'graphql-iso-date';
import GraphQLJSON from 'graphql-type-json';

import { notNullOrUndefined } from '../../../../shared/shared-utils';
import { ConfigModule } from '../../config/config.module';
import { ConfigService } from '../../config/config.service';
import { I18nModule } from '../../i18n/i18n.module';
import { I18nService } from '../../i18n/i18n.service';
import { TranslateErrorExtension } from '../middleware/translate-errors-extension';

import { generateListOptions } from './generate-list-options';
import { addGraphQLCustomFields } from './graphql-custom-fields';

export interface GraphQLApiOptions {
    typePaths: string[];
    apiPath: string;
    // tslint:disable-next-line:ban-types
    resolverModule: Function;
}

/**
 * Dynamically generates a GraphQLModule according to the given config options.
 */
export function configureGraphQLModule(
    getOptions: (configService: ConfigService) => GraphQLApiOptions,
): DynamicModule {
    return GraphQLModule.forRootAsync({
        useFactory: (
            configService: ConfigService,
            i18nService: I18nService,
            typesLoader: GraphQLTypesLoader,
        ) => {
            return createGraphQLOptions(i18nService, configService, typesLoader, getOptions(configService));
        },
        inject: [ConfigService, I18nService, GraphQLTypesLoader],
        imports: [ConfigModule, I18nModule],
    });
}

function createGraphQLOptions(
    i18nService: I18nService,
    configService: ConfigService,
    typesLoader: GraphQLTypesLoader,
    options: GraphQLApiOptions,
): GqlModuleOptions {
    // Prevent `Type "Node" is missing a "resolveType" resolver.` warnings.
    // See https://github.com/apollographql/apollo-server/issues/1075
    const dummyResolveType = {
        __resolveType() {
            return null;
        },
    };

    return {
        path: '/' + options.apiPath,
        typeDefs: createTypeDefs(),
        include: [options.resolverModule],
        resolvers: {
            JSON: GraphQLJSON,
            DateTime: GraphQLDateTime,
            Node: dummyResolveType,
            PaginatedList: dummyResolveType,
            Upload: GraphQLUpload || dummyResolveType,
        },
        uploads: {
            maxFileSize: configService.assetOptions.uploadMaxFileSize,
        },
        playground: true,
        debug: true,
        context: (req: any) => req,
        extensions: [() => new TranslateErrorExtension(i18nService)],
        // This is handled by the Express cors plugin
        cors: false,
    };

    /**
     * Generates the server's GraphQL schema by combining:
     * 1. the default schema as defined in the source .graphql files specified by `typePaths`
     * 2. any custom fields defined in the config
     * 3. any schema extensions defined by plugins
     */
    function createTypeDefs(): string {
        const customFields = configService.customFields;
        const typeDefs = typesLoader.mergeTypesByPaths(...options.typePaths);
        let schema = generateListOptions(typeDefs);
        schema = addGraphQLCustomFields(schema, customFields);
        const pluginTypes = configService.plugins
            .map(p => (p.defineGraphQlTypes ? p.defineGraphQlTypes() : undefined))
            .filter(notNullOrUndefined);
        for (const types of pluginTypes) {
            schema = extendSchema(schema, types);
        }
        return printSchema(schema);
    }
}
