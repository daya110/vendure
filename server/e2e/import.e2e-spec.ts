import gql from 'graphql-tag';
import path from 'path';

import { TEST_SETUP_TIMEOUT_MS } from './config/test-config';
import { TestClient } from './test-client';
import { TestServer } from './test-server';

describe('Import resolver', () => {
    const client = new TestClient();
    const server = new TestServer();

    beforeAll(async () => {
        const token = await server.init({
            productCount: 0,
            customerCount: 0,
        });
        await client.init();
    }, TEST_SETUP_TIMEOUT_MS);

    afterAll(async () => {
        await server.destroy();
    });

    it('imports products', async () => {
        // TODO: waste a few more hours actually fixing this for real
        // Forgive me this abomination of a work-around.
        // On the inital run (as in CI), when the sqlite db has just been populated,
        // this test will fail due to an "out of memory" exception originating from
        // SqljsQueryRunner.ts:79:22, which is part of the findOne() operation on the
        // Session repository called from the AuthService.validateSession() method.
        // After several hours of fruitless hunting, I did what any desperate JavaScript
        // developer would do, and threw in a setTimeout. Which of course "works"...
        const timeout = process.env.CI ? 2000 : 1000;
        await new Promise(resolve => {
            setTimeout(resolve, timeout);
        });

        const csvFile = path.join(__dirname, 'fixtures', 'product-import.csv');
        const result = await client.importProducts(csvFile);

        expect(result.importProducts.errors).toEqual([
            'Invalid Record Length: header length is 12, got 1 on line 8',
        ]);
        expect(result.importProducts.imported).toBe(4);
        expect(result.importProducts.processed).toBe(4);

        const productResult = await client.query(
            gql`
                query GetProducts($options: ProductListOptions) {
                    products(options: $options) {
                        totalItems
                        items {
                            id
                            name
                            slug
                            description
                            featuredAsset {
                                id
                                name
                            }
                            assets {
                                id
                                name
                            }
                            optionGroups {
                                id
                                code
                                name
                            }
                            facetValues {
                                id
                                name
                                facet {
                                    id
                                    name
                                }
                            }
                            variants {
                                id
                                name
                                sku
                                price
                                taxCategory {
                                    id
                                    name
                                }
                                options {
                                    id
                                    code
                                }
                                assets {
                                    id
                                    name
                                }
                                featuredAsset {
                                    id
                                    name
                                }
                                facetValues {
                                    id
                                    code
                                    name
                                    facet {
                                        id
                                        name
                                    }
                                }
                            }
                        }
                    }
                }
            `,
            {
                options: {},
            },
        );

        expect(productResult.products.totalItems).toBe(4);
        expect(productResult.products.items).toMatchSnapshot();
        return;
    }, 10000);
});
