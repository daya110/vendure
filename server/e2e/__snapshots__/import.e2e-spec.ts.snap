// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import resolver imports products 1`] = `
Array [
  Object {
    "assets": Array [
      Object {
        "id": "T_6",
        "name": "pps1.jpg",
      },
      Object {
        "id": "T_7",
        "name": "pps2.jpg",
      },
    ],
    "description": "A great device for stretching paper.",
    "featuredAsset": Object {
      "id": "T_6",
      "name": "pps1.jpg",
    },
    "id": "T_1",
    "name": "Perfect Paper Stretcher",
    "optionGroups": Array [
      Object {
        "code": "perfect-paper-stretcher-size",
        "id": "T_2",
        "name": "size",
      },
    ],
    "slug": "perfect-paper-stretcher",
    "variants": Array [
      Object {
        "assets": Array [],
        "featuredAsset": null,
        "id": "T_1",
        "name": "Perfect Paper Stretcher Half Imperial",
        "options": Array [
          Object {
            "code": "half-imperial",
            "id": "T_3",
          },
        ],
        "price": 4530,
        "sku": "PPS12",
        "taxCategory": Object {
          "id": "T_1",
          "name": "Standard Tax",
        },
      },
      Object {
        "assets": Array [],
        "featuredAsset": null,
        "id": "T_2",
        "name": "Perfect Paper Stretcher Quarter Imperial",
        "options": Array [
          Object {
            "code": "quarter-imperial",
            "id": "T_4",
          },
        ],
        "price": 3250,
        "sku": "PPS14",
        "taxCategory": Object {
          "id": "T_1",
          "name": "Standard Tax",
        },
      },
      Object {
        "assets": Array [],
        "featuredAsset": null,
        "id": "T_3",
        "name": "Perfect Paper Stretcher Full Imperial",
        "options": Array [
          Object {
            "code": "full-imperial",
            "id": "T_5",
          },
        ],
        "price": 5950,
        "sku": "PPSF",
        "taxCategory": Object {
          "id": "T_1",
          "name": "Standard Tax",
        },
      },
    ],
  },
  Object {
    "assets": Array [],
    "description": "Mabef description",
    "featuredAsset": null,
    "id": "T_2",
    "name": "Mabef M/02 Studio Easel",
    "optionGroups": Array [],
    "slug": "mabef-m02-studio-easel",
    "variants": Array [
      Object {
        "assets": Array [],
        "featuredAsset": null,
        "id": "T_4",
        "name": "Mabef M/02 Studio Easel",
        "options": Array [],
        "price": 91070,
        "sku": "M02",
        "taxCategory": Object {
          "id": "T_1",
          "name": "Standard Tax",
        },
      },
    ],
  },
  Object {
    "assets": Array [],
    "description": "Really mega pencils",
    "featuredAsset": null,
    "id": "T_3",
    "name": "Giotto Mega Pencils",
    "optionGroups": Array [
      Object {
        "code": "giotto-mega-pencils-box-size",
        "id": "T_3",
        "name": "box size",
      },
    ],
    "slug": "giotto-mega-pencils",
    "variants": Array [
      Object {
        "assets": Array [
          Object {
            "id": "T_8",
            "name": "box-of-8.jpg",
          },
        ],
        "featuredAsset": Object {
          "id": "T_8",
          "name": "box-of-8.jpg",
        },
        "id": "T_5",
        "name": "Giotto Mega Pencils Box of 8",
        "options": Array [
          Object {
            "code": "box-of-8",
            "id": "T_6",
          },
        ],
        "price": 416,
        "sku": "225400",
        "taxCategory": Object {
          "id": "T_1",
          "name": "Standard Tax",
        },
      },
      Object {
        "assets": Array [
          Object {
            "id": "T_9",
            "name": "box-of-12.jpg",
          },
        ],
        "featuredAsset": Object {
          "id": "T_9",
          "name": "box-of-12.jpg",
        },
        "id": "T_6",
        "name": "Giotto Mega Pencils Box of 12",
        "options": Array [
          Object {
            "code": "box-of-12",
            "id": "T_7",
          },
        ],
        "price": 624,
        "sku": "225600",
        "taxCategory": Object {
          "id": "T_1",
          "name": "Standard Tax",
        },
      },
    ],
  },
  Object {
    "assets": Array [],
    "description": "Keeps the paint off the clothes",
    "featuredAsset": null,
    "id": "T_4",
    "name": "Artists Smock",
    "optionGroups": Array [
      Object {
        "code": "artists-smock-size",
        "id": "T_4",
        "name": "size",
      },
      Object {
        "code": "artists-smock-colour",
        "id": "T_5",
        "name": "colour",
      },
    ],
    "slug": "artists-smock",
    "variants": Array [
      Object {
        "assets": Array [],
        "featuredAsset": null,
        "id": "T_7",
        "name": "Artists Smock small beige",
        "options": Array [
          Object {
            "code": "small",
            "id": "T_8",
          },
          Object {
            "code": "beige",
            "id": "T_10",
          },
        ],
        "price": 1199,
        "sku": "10112",
        "taxCategory": Object {
          "id": "T_2",
          "name": "Reduced Tax",
        },
      },
      Object {
        "assets": Array [],
        "featuredAsset": null,
        "id": "T_8",
        "name": "Artists Smock large beige",
        "options": Array [
          Object {
            "code": "large",
            "id": "T_9",
          },
          Object {
            "code": "beige",
            "id": "T_10",
          },
        ],
        "price": 1199,
        "sku": "10113",
        "taxCategory": Object {
          "id": "T_2",
          "name": "Reduced Tax",
        },
      },
      Object {
        "assets": Array [],
        "featuredAsset": null,
        "id": "T_9",
        "name": "Artists Smock small navy",
        "options": Array [
          Object {
            "code": "small",
            "id": "T_8",
          },
          Object {
            "code": "navy",
            "id": "T_11",
          },
        ],
        "price": 1199,
        "sku": "10114",
        "taxCategory": Object {
          "id": "T_2",
          "name": "Reduced Tax",
        },
      },
      Object {
        "assets": Array [],
        "featuredAsset": null,
        "id": "T_10",
        "name": "Artists Smock large navy",
        "options": Array [
          Object {
            "code": "large",
            "id": "T_9",
          },
          Object {
            "code": "navy",
            "id": "T_11",
          },
        ],
        "price": 1199,
        "sku": "10115",
        "taxCategory": Object {
          "id": "T_2",
          "name": "Reduced Tax",
        },
      },
    ],
  },
]
`;
